### PR TITLE
feature(format): refactor output format

### LINKED ISSUE #5433: Tracking support output format
**Summary**

Description for this feature.

- Support format output in clickhouse handler
- support ndjson/csv/parquet/tsv output formats
- stateless tests to cover output format

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR:

1.  Make clickhouse handler support output format.
2. Make clickhouse handler support post SQL in body.
3. Introduce parquet/csv/ndjson output format

## Changelog

- New Feature
 

## Related Issues

Fixes #5433

### CODE REVIEW

**Review by BohuTANG — APPROVED** [review]
Nice job!
I think this is useful for the result cache cc @youngsofun

**Comment by BohuTANG** [comment]
The unit-test is skipped because the disk space is full:
```
[test_unit](https://github.com/datafuselabs/databend/runs/6482164666?check_suite_focus=true)
System.IO.IOException: No space left on device : '/home/runner/runners/2.291.1/_diag/Worker_20220518-030049-utc.log'
   at System.IO.RandomAccess.WriteAtOffset(SafeFileHandle handle, ReadOnlySpan`1 buffer, Int64 fileOffset)
   at System.IO.Strategies.BufferedFileStreamStrategy.FlushWrite()
   at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
   at System.Diagnostics.TextWriterTraceListener.Flush()
   at GitHub.Runner.Common.HostTraceListener.WriteHeader(String source, TraceEventType eventType, Int32 id)
   at GitHub.Runner.Common.HostTraceListener.TraceEvent(TraceEventCache eventCache, String source, TraceEventType eventType, Int32 id, String message)
   at System.Diagnostics.TraceSource.TraceEvent(TraceEventType eventType, Int32 id, String message)
   at GitHub.Runner.Worker.Worker.RunAsync(String pipeIn, String pipeOut)
   at GitHub.Runner.Worker.Program.MainAsync(IHostContext context, String[] args)
System.IO.IOException: No space left on device : '/home/runner/runners/2.291.1/_diag/Worker_20220518-030049-utc.log'
   at System.IO.RandomAccess.WriteAtOffset(SafeFileHandle handle, ReadOnlySpan`1 buffer, Int64 fileOffset)
   at System.IO.Strategies.BufferedFileStreamStrategy.FlushWrite()
   at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
   at System.Diagnostics.TextWriterTraceListener.Flush()
   at GitHub.Runner.Common.HostTraceListener.WriteHeader(String source, TraceEventType eventType, Int32 id)
   at GitHub.Runner.Common.HostTraceListener.TraceEvent(TraceEventCache eventCache, String source, TraceEventType eventType, Int32 id, String message)
   at System.Diagnostics.TraceSource.TraceEvent(TraceEventType eventType, Int32 id, String message)
   at GitHub.Runner.Common.Tracing.Error(Exception exception)
   at GitHub.Runner.Worker.Program.MainAsync(IHostContext context, String[] args)
Unhandled exception. System.IO.IOException: No space left on device : '/home/runner/runners/2.291.1/_diag/Worker_20220518-030049-utc.log'
   at System.IO.RandomAccess.WriteAtOffset(SafeFileHandle handle, ReadOnlySpan`1 buffer, Int64 fileOffset)
   at System.IO.Strategies.BufferedFileStreamStrategy.FlushWrite()
   at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
   at System.Diagnostics.TextWriterTraceListener.Flush()
   at System.Diagnostics.TraceSource.Flush()
   at GitHub.Runner.Common.TraceManager.Dispose(Boolean disposing)
   at GitHub.Runner.Common.TraceManager.Dispose()
   at GitHub.Runner.Common.HostContext.Dispose(Boolean disposing)
   at GitHub.Runner.Common.HostContext.Dispose()
   at GitHub.Runner.Worker.Program.Main(String[] args)
```

**Comment by BohuTANG** [comment]
@sundy-li Seems some tests break the unit-test in this patch. There is no space left for the CI, so the unit-test is skipped every time.

image

**Comment by BohuTANG** [comment]
Oops, it's not related to this patch, another PR has the same issue too:
https://github.com/datafuselabs/databend/actions/runs/2343011703

Ping @Xuanwo for help

File: query/src/servers/http/clickhouse_handler.rs [hunk 1]
```diff
@@ -181,38 +151,34 @@ pub async fn clickhouse_handler_post(
...
 
-        let input_stream = match format {
-            Format::NDJson => build_ndjson_stream(&plan, body).await.map_err(BadRequest)?,
-        };
-        (plan, Some(input_stream))
+        (plan, format, None)
     } else {
-        // Other sql
-        let body = body.into_string().await.map_err(BadRequest)?;
-        let sql = format!("{}\n{}", sql, body);
```
Comment (youngsofun) [inline query/src/servers/http/clickhouse_handler.rs:203]: @sundy-li  

no longer support `let sql = format!("{}\n{}", sql, body);` ?

Comment (youngsofun) [reply query/src/servers/http/clickhouse_handler.rs:203]: and assume insert query always let SQL part in query param and  DATA part in body?

File: common/datavalues/src/types/serializations/string.rs [hunk 2]
```diff
@@ -47,6 +47,19 @@ impl TypeSerializer for StringSerializer {
...
 
+    fn serialize_column_quoted(
+        &self,
+        column: &ColumnRef,
+        _format: &FormatSettings,
+    ) -> Result<Vec<String>> {
+        let column: &StringColumn = Series::check_get(column)?;
+        let result: Vec<String> = column
+            .iter()
+            .map(|v| format!("{:?}", String::from_utf8_lossy(v)))
```
Comment (youngsofun) [inline common/datavalues/src/types/serializations/string.rs:58]: need escape, but we can do it later.

File: query/src/servers/http/clickhouse_handler.rs (hunk 1 above)
Comment (sundy-li) [reply query/src/servers/http/clickhouse_handler.rs:203]: I think it's reasonable, if the sql is empty then the body is the whole sql.

**Comment by Xuanwo** [comment]
> Oops, it's not related to this patch, another PR has the same issue too: https://github.com/datafuselabs/databend/actions/runs/2343011703
> 
> Ping @Xuanwo @everpcpc for help

Maybe related to our rust cache, I have reset it to see whether it works.

File: common/ast/tests/it/parser.rs [hunk 3]
```diff
@@ -186,6 +186,7 @@ fn test_query() {
     let mut mint = Mint::new("tests/it/testdata");
     let mut file = mint.new_goldenfile("query.txt").unwrap();
     let cases = &[
+        r#"select * from a limit 3 offset 4 format csv"#,
```
Comment (youngsofun) [inline common/ast/tests/it/parser.rs:189]: @sundy-li  where supposed to use the format in this SQL?

**Review by youngsofun — APPROVED** [review]

File: common/ast/tests/it/parser.rs (hunk 3 above)
Comment (sundy-li) [reply common/ast/tests/it/parser.rs:189]: At the end of the sql.

**Comment by sundy-li** [comment]
Now unit tests have some bugs, I'll fix them later.

**Comment by Xuanwo** [comment]
Is test_unit able to run without hitting `no space left` error?

**Pull request merged**